src/codecs/openexr.rs

```diff
@@ -0,0 +1,495 @@
+
+    fn dimensions(&self) -> (u32, u32) {
+        let size = self.selected_exr_header().shared_attributes.display_window.size;
+        (size.width() as u32, size.height() as u32)
+    }
+
+    fn color_type(&self) -> ColorType {
+        let returns_alpha = self.alpha_preference.unwrap_or(self.alpha_present_in_file);
+        if returns_alpha { ColorType::Rgba32F } else { ColorType::Rgb32F }
+    }
+
+    fn original_color_type(&self) -> ExtendedColorType {
+        if self.alpha_present_in_file { ExtendedColorType::Rgba32F } else { ExtendedColorType::Rgb32F }
+    }
+
+    /// Use `read_image` instead if possible,
+    /// as this method creates a whole new buffer just to contain the entire image.
+    fn into_reader(self) -> ImageResult<Self::Reader> {
+        Ok(Cursor::new(decoder_to_vec(self)?))
+    }
+
+    fn scanline_bytes(&self) -> u64 {
+        // we cannot always read individual scan lines for every file,
+        // as the tiles or lines in the file could be in random or reversed order.
+        // therefore we currently read all lines at once
+        // Todo: optimize for specific exr.line_order?
+       self.total_bytes()
+    }
+
+    // reads with or without alpha, depending on `self.alpha_preference` and `self.alpha_present_in_file`
+    fn read_image_with_progress<F: Fn(Progress)>(self, unaligned_bytes: &mut [u8], progress_callback: F) -> ImageResult<()> {
+        let blocks_in_header = self.selected_exr_header().chunk_count as u64;
+        let channel_count = self.color_type().channel_count() as usize;
+
+        let display_window = self.selected_exr_header().shared_attributes.display_window;
+        let data_window_offset = self.selected_exr_header().own_attributes.layer_position - display_window.position;
+
+        let result = read()
+            .no_deep_data().largest_resolution_level()
+            .rgba_channels(
+                move |_size, _channels| {
+                    vec![0_f32; display_window.size.area() * channel_count]
+                },
+
+                move |buffer, index_in_data_window, (r,g,b,a_or_1): (f32,f32,f32,f32)| {
+                    let index_in_display_window = index_in_data_window.to_i32() + data_window_offset;
+
+                    // only keep pixels inside the data window
+                    // TODO filter chunks based on this
+                    if index_in_display_window.x() >= 0 && index_in_display_window.y() >= 0
+                        && index_in_display_window.x() < display_window.size.width() as i32
+                        && index_in_display_window.y() < display_window.size.height() as i32
+                    {
+                        let index_in_display_window = index_in_display_window.to_usize("index bug").unwrap();
+                        let first_f32_index = index_in_display_window.flat_index_for_size(display_window.size);
+
+                        buffer[first_f32_index * channel_count .. (first_f32_index + 1) * channel_count]
```

Comment:
In the decoder, this indexing operation will panic when the buffer has an invalid size. The size should probably be checked and return an error in case it is invalid, right?